mteb/evaluation/evaluators/RetrievalEvaluator.py

```diff
@@ -221,6 +243,171 @@ def is_dres_compatible(model):
     return True
 
 
+def is_cross_encoder_compatible(model):
+    op = getattr(model, "predict", None)
+    if not (callable(op)):
+        return False
+    return True
+
+
+class Reranker:
+    """
+    This class provides support for reranker (or cross-encoder) models that encoder query and document at the same time (typically with attention).
+        Some notable examples include MonoBERT, MonoT5, RankLlama, etc.
+        Note: you must provide either a first stage model or the results of the first stage model so that it can rerank.
+        You also must extend this class and provide an `__init__` function that loads your model and a `rerank` function that reranks a batch.
+    """
+
+    def __init__(
+        self,
+        model: str,
+        first_stage: str | CrossEncoder | None = None,
```

Comment:
Like I said earlier, I'm pretty unfamiliar with Protocols in Python.  Can I use this syntax or is there a place to learn this syntax?

Reply:
I believe the [pep](https://peps.python.org/pep-0544/) is a fairly good introduction to it (though that might just be my preference)

but essentially the specification above would work though you would have to add a decorator to make it runtime checkable:

```
@runtime_checkable
class CrossEncoder(Protocol):
  predict(sentences: list[str]) -> list[float]: 
    ...
```

A few more examples (just to get you familiar with the syntax):

For the encoder:
```
@runtime_checkable
class Encoder(Protocol):
  encode(sentences: list[str], prompt: str, **kwargs) -> np.ndarray | torch.Tensor: 
    ...
```
Though **kwargs would have to be expanded to the keys we actually pass it.

for the extended encoder it would be something like this:
```
@runtime_checkable
class EncoderWithQueryCorpusEncode(Encoder):  # is also an Encoder
  encode_curpus(**kwargs) -> np.ndarray | torch.Tensor: 
    ...
  encode_queries(**kwargs) -> np.ndarray | torch.Tensor: 
    ...
```
Again **kwargs would need to be specified more.

Reply:
> but essentially the specification above would work though you would have to add a decorator to make it runtime checkable:

I can add this as is the code, or we can make it in another PR.  It may take me some time to wrap my head around the syntax - though I love the idea!

Reply:
Might be another PR (I have created an issue on it https://github.com/embeddings-benchmark/mteb/issues/464)